packages/runtime/src/next/api-handler/handlers/bypass-token.ts

```diff
@@ -0,0 +1,107 @@
+  | [request: NextRequest, context: Context, params: { apiKey: string }]
+  | [req: NextApiRequest, res: NextApiResponse<BypassTokenResponse>, params: { apiKey: string }]
+
+const routeHandlerPattern = [P.instanceOf(Request), P.any, P.any] as const
+const apiRoutePattern = [P.any, P.any, P.any] as const
+
+export default async function bypassTokenHandler(
+  request: NextRequest,
+  context: Context,
+  { apiKey }: { apiKey: string },
+): Promise<NextResponse<BypassTokenResponse>>
+export default async function bypassTokenHandler(
+  req: NextApiRequest,
+  res: NextApiResponse<BypassTokenResponse>,
+  { apiKey }: { apiKey: string },
+): Promise<void>
+export default async function bypassTokenHandler(
+  ...args: BypassTokenHandlerArgs
+): Promise<NextResponse<BypassTokenResponse> | void> {
+  return match(args)
+    .with(routeHandlerPattern, args => bypassTokenRouteHandler(...args))
+    .with(apiRoutePattern, args => bypassTokenApiRouteHandler(...args))
+    .exhaustive()
+}
+
+async function bypassTokenRouteHandler(
+  request: NextRequest,
+  _context: Context,
+  { apiKey }: { apiKey: string },
+): Promise<NextResponse<BypassTokenResponse>> {
+  const secret =
+    request.nextUrl.searchParams.get(SECRET_HEADER) ?? request.headers.get(SECRET_HEADER)
+
+  if (secret !== apiKey) return new NextResponse('Unauthorized', { status: 401 })
+
+  const draft = await draftMode()
+  const cookieStore = await cookies()
+
+  draft.enable()
+  const prerenderBypassToken = cookieStore.get(PRERENDER_BYPASS_COOKIE)
+
+  if (prerenderBypassToken?.value == null) {
+    return new NextResponse('Failed to retrieve token', { status: 500 })
+  }
+
+  return NextResponse.json({
+    cookies: [
+      { name: PRERENDER_BYPASS_COOKIE, value: prerenderBypassToken.value },
+      {
+        name: DRAFT_DATA_COOKIE,
+        value: JSON.stringify({ makeswift: true, siteVersion: 'Working' }),
+      },
+    ],
+  })
+}
+
+async function bypassTokenApiRouteHandler(
+  req: NextApiRequest,
+  res: NextApiResponse<BypassTokenResponse>,
+  { apiKey }: { apiKey: string },
+): Promise<void> {
+  const secret = req.query[SECRET_HEADER]
+
+  if (secret !== apiKey) return res.status(401).send('Unauthorized')
+
+  const setCookie = res
+    .setPreviewData({ makeswift: true, siteVersion: 'Working' })
+    .getHeader('set-cookie')
+
+  res.removeHeader('set-cookie')
+
+  const parsedCookies = parseSetCookie(Array.isArray(setCookie) ? setCookie : '')
+
+  const prerenderBypassCookie = parsedCookies.find(c => c.name === PRERENDER_BYPASS_COOKIE)
+  const previewDataCookie = parsedCookies.find(c => c.name === PREVIEW_DATA_COOKIE)
+
+  if (prerenderBypassCookie?.value == null || previewDataCookie?.value == null) {
+    return res.status(500).send('Failed to retrieve token')
+  }
+
+  return res.json({
+    cookies: [
+      { name: PRERENDER_BYPASS_COOKIE, value: prerenderBypassCookie.value },
+      { name: PREVIEW_DATA_COOKIE, value: previewDataCookie.value },
```

Comment:
Should we use the cookie `.name` fields here? Seems more semantically appropriate